src/udev/udevd.c

```diff
@@ -545,25 +555,23 @@ static void worker_spawn(Manager *manager, struct event *event) {
 }
 
 static void event_run(Manager *manager, struct event *event) {
-        struct worker *worker;
-        Iterator i;
-
         assert(manager);
         assert(event);
-
-        HASHMAP_FOREACH(worker, manager->workers, i) {
+retry:
+        if (!LIST_IS_EMPTY(manager->idle_workers)) {
+                struct worker *worker = manager->idle_workers;
                 ssize_t count;
 
-                if (worker->state != WORKER_IDLE)
-                        continue;
+                assert(worker->state == WORKER_IDLE);
```

Comment:
@martinpitt To be honest I haven't tested performance yet (have to set up some hw for it). If i get too it, I'll whip up some A/B testing tomorrow.
Yeah, the assert might be a bit over the top, i could just go with an if, an error log, and goto retry.

I'll test performance first, since that was pretty much the whole reason for it.